event/consumer.go

```diff
@@ -2,79 +2,127 @@ package event
-func Consume(ctx context.Context, messageConsumer kafka.IConsumerGroup, handler Handler, cfg *config.Config) {
+func (consumer *Consumer) Consume(
+	ctx context.Context,
+	messageConsumer MessageConsumer,
+	handler Handler,
+	cfg *config.Config) {
+
+	go func() {
+		defer close(consumer.closed)
 
-	// consume loop, to be executed by each worker
-	var consume = func(workerID int) {
+		batch := NewBatch(cfg.BatchSize)
+		// Wait a batch full of messages.
+		// If we do not get any messages for a time, just process the messages already in the batch.
 		for {
 			select {
-			case message, ok := <-messageConsumer.Channels().Upstream:
-				if !ok {
-					log.Info(ctx, "closing event consumer loop because upstream channel is closed", log.Data{"worker_id": workerID})
-					return
+			case msg := <-messageConsumer.Channels().Upstream:
+				ctx := context.Background()
+
+				AddMessageToBatch(ctx, batch, msg, handler)
+				msg.Release()
+
+			case <-time.After(cfg.BatchWaitTime):
+				if batch.IsEmpty() {
+					continue
 				}
-				messageCtx := context.Background()
-				processMessage(messageCtx, message, handler, cfg)
-				message.Release()
-			case <-messageConsumer.Channels().Closer:
-				log.Info(ctx, "closing event consumer loop because closer channel is closed", log.Data{"worker_id": workerID})
+
+				ctx := context.Background()
+
+				log.Event(ctx, "batch wait time reached. proceeding with batch", log.INFO, log.Data{"batchsize": batch.Size()})
+				ProcessBatch(ctx, handler, batch)
+
+			case eventClose := <-consumer.closing:
+				log.Event(eventClose.ctx, "closing event consumer loop", log.INFO)
+				close(consumer.closing)
 				return
 			}
 		}
-	}
+	}()
+}
 
-	// workers to consume messages in parallel
-	for w := 1; w <= cfg.KafkaNumWorkers; w++ {
-		go consume(w)
+// AddMessageToBatch will attempt to add the message to the batch and determine if it should be processed.
+func AddMessageToBatch(ctx context.Context, batch *Batch, msg kafka.Message, handler Handler) {
+	log.Info(ctx, "add message to batch starts", log.Data{"add message to batch starts time": time.Now()})
+	batch.Add(ctx, msg)
+	if batch.IsFull() {
+		log.Event(ctx, "batch is full - processing batch", log.INFO, log.Data{"batchsize": batch.Size()})
+		ProcessBatch(ctx, handler, batch)
 	}
 }
 
-// processMessage unmarshals the provided kafka message into an event and calls the handler.
-// After the message is handled, it is committed.
-func processMessage(ctx context.Context, message kafka.Message, handler Handler, cfg *config.Config) {
-
-	// unmarshal - commit on failure (consuming the message again would result in the same error)
-	event, err := unmarshal(message)
+// ProcessBatch will attempt to handle and commit the batch, or shutdown if something goes horribly wrong.
+func ProcessBatch(ctx context.Context, handler Handler, batch *Batch) {
+	log.Info(ctx, "process batch starts", log.Data{
+		"process batch starts time": time.Now(),
```

Comment:
Please remove `"process batch starts time"` from log.Data; the log library creates a timestamp for a log event (info, error etc.). Also when choosing a key name, please use `snake_case` as this will be displayed as JSON.